File: sdks/php/src/Webhooks.php
<?php

declare(strict_types=1);

namespace Pushfy;

/**
 * Verifies the authenticity of an incoming Pushfy webhook.
 *
 * Signature header differs by product:
 *   - Messaging status   -> X-Pushfy-Signature: sha256=<hex>   (scheme: 'prefixed')
 *   - Push Notifications  -> X-Push-Signature:   sha256=<hex>   (scheme: 'prefixed')
 *   - Conversational AI   -> X-PA-Signature:      <hex>          (scheme: 'raw')
 *
 * Always pass the RAW request body (the exact bytes received), not a
 * re-serialized array — re-serialization changes the signature.
 */
final class Webhooks
{
    public const SCHEME_PREFIXED = 'prefixed';
    public const SCHEME_RAW = 'raw';

    /**
     * @param string $payload    Raw request body (exact bytes).
     * @param string $signature  Value of the signature header.
     * @param string $secret     Your webhook secret.
     * @param string $scheme     'prefixed' (sha256=<hex>) or 'raw' (<hex>).
     * @return bool  true when the signature is valid.
     */
    public static function verify(string $payload, ?string $signature, string $secret, string $scheme = self::SCHEME_PREFIXED): bool
    {
        if ($signature === null || $signature === '' || $secret === '') {
            return false;
        }

        $hex = hash_hmac('sha256', $payload, $secret);
        $expected = $scheme === self::SCHEME_RAW ? $hex : 'sha256=' . $hex;

        // hash_equals is constant-time and length-safe.
        return hash_equals($expected, $signature);
    }

    /** Messaging status/DLR webhook: X-Pushfy-Signature (sha256=<hex>). */
    public static function messaging(string $payload, ?string $signature, string $secret): bool
    {
        return self::verify($payload, $signature, $secret, self::SCHEME_PREFIXED);
    }

    /** Push Notifications webhook: X-Push-Signature (sha256=<hex>). */
    public static function push(string $payload, ?string $signature, string $secret): bool
    {
        return self::verify($payload, $signature, $secret, self::SCHEME_PREFIXED);
    }

    /** Conversational AI (PushAgent) webhook: X-PA-Signature (raw hex). */
    public static function conversations(string $payload, ?string $signature, string $secret): bool
    {
        return self::verify($payload, $signature, $secret, self::SCHEME_RAW);
    }
}

File: sdks/go/webhooks.go
package pushfy

import (
	"crypto/hmac"
	"crypto/sha256"
	"encoding/hex"
)

// Scheme selects how the incoming signature header is formatted.
type Scheme int

const (
	// SchemePrefixed expects "sha256=<hex>" (Messaging status/DLR and Push
	// Notifications: X-Pushfy-Signature / X-Push-Signature).
	SchemePrefixed Scheme = iota
	// SchemeRaw expects a bare "<hex>" (Conversational AI: X-PA-Signature).
	SchemeRaw
)

// Verify checks the authenticity of an incoming Pushfy webhook.
//
// Always pass the RAW request body (the exact bytes received), not a
// re-serialized object — re-serialization changes the signature. The comparison
// is constant-time (crypto/hmac.Equal).
func Verify(payload []byte, signature, secret string, scheme Scheme) bool {
	if signature == "" || secret == "" {
		return false
	}
	mac := hmac.New(sha256.New, []byte(secret))
	mac.Write(payload)
	sum := hex.EncodeToString(mac.Sum(nil))

	expected := sum
	if scheme == SchemePrefixed {
		expected = "sha256=" + sum
	}
	return hmac.Equal([]byte(expected), []byte(signature))
}

// VerifyMessaging verifies a Messaging status/DLR webhook (X-Pushfy-Signature,
// "sha256=" prefixed).
func VerifyMessaging(payload []byte, signature, secret string) bool {
	return Verify(payload, signature, secret, SchemePrefixed)
}

// VerifyPush verifies a Push Notifications webhook (X-Push-Signature, "sha256="
// prefixed).
func VerifyPush(payload []byte, signature, secret string) bool {
	return Verify(payload, signature, secret, SchemePrefixed)
}

// VerifyConversations verifies a Conversational AI / PushAgent webhook
// (X-PA-Signature, raw hex).
func VerifyConversations(payload []byte, signature, secret string) bool {
	return Verify(payload, signature, secret, SchemeRaw)
}
